// purs-loader/lib/index.js
'use strict';

var debug_ = require('debug');

var debug = debug_('purs-loader');

var debugVerbose = debug_('purs-loader:verbose');

var loaderUtils = require('loader-utils');

var Promise = require('bluebird');

var path = require('path');

var globby = require('globby');

var PsModuleMap = require('./purs-module-map');

var compile = require('./compile');

var bundle = require('./bundle');

var ide = require('./ide');

var toJavaScript = require('./to-javascript');

var sourceMaps = require('./source-maps');

var dargs = require('./dargs');

var spawn = require('cross-spawn').sync;

var eol = require('os').EOL;

var CACHE_VAR = {
  rebuild: false,
  deferred: [],
  bundleModules: [],
  ideServer: null,
  psModuleMap: null,
  warnings: [],
  errors: [],
  compilationStarted: false,
  compilationFinished: false,
  compilationFailed: false,
  installed: false,
  srcOption: []
};

const matchErrorsSeparator = /Error \d+ of \d+:/;
const matchLocation = /at (.+\.purs) line (\d+), column (\d+) - line (\d+), column (\d+)/;
const matchModuleName = /in module ((?:\w+\.)*\w+)/;
const matchMissingModuleName = /Module ((?:\w+\.)*\w+) was not found/;
const matchMissingImportFromModuleName = /Cannot import value \w+ from module ((?:\w+\.)*\w+)/;

const pscErrorsSeparator = /\n(?=Error)/;
const splitPscErrors = pscMessage =>
  pscMessage.split(pscErrorsSeparator);

const repeat = (value, times) => {
  if (times <= 0) return [];
  const repeated = [];
  for (let n = 0; n < times; n += 1) {
    repeated.push(value);
  }
  return value;
};

const diffModuleNames = (from, target, parts) => {
  if (!from.length) return parts.concat(target);
  if (!target.length) return parts.concat(repeat("..", from.length));
  const [head_from, ...tail_from] = from;
  const [head_target, ...tail_target] = target;
  return head_from === head_target
    ? diffModuleNames(tail_from, tail_target, parts)
    : parts.concat(repeat("..", from.length), target);
};

const resolveFilename = ({ base, from, target, ext = ".purs" }) => {
  const parts = diffModuleNames(from.split("."), target.split("."), []);
  return parts.length
    ? path.resolve(base, path.join(...parts) + ext)
    : path.join(path.dirname(base),
        path.basename(base, path.extname(base)) + ext);
};

module.exports = function purescriptLoader(source, map) {

  // BRK
  // if (!/.psc-package/.test(this.resourcePath)) debugger;

  var _this = this;

  this.cacheable && this.cacheable();

  var webpackContext = this.options && this.options.context || this.rootContext;

  var callback = this.async();

  var loaderOptions = loaderUtils.getOptions(this) || {};
  var loaderSrc = loaderOptions.src || [
    path.join('src', '**', '*.purs')
  ];

  var srcOption = function (pscPackage) {
    var srcPath = path.join('src', '**', '*.purs');

    var bowerPath = path.join('bower_components', 'purescript-*', 'src', '**', '*.purs');

    if (CACHE_VAR.srcOption.length > 0) {
      return CACHE_VAR.srcOption;
    } else if (pscPackage) {
      var pscPackageCommand = 'psc-package';

      var pscPackageArgs = ['sources'];

      var loaderSrc = loaderOptions.src || [srcPath];

      debug('psc-package %s %o', pscPackageCommand, pscPackageArgs);

      var cmd = spawn(pscPackageCommand, pscPackageArgs);

      if (cmd.error) {
        throw new Error(cmd.error);
      } else if (cmd.status !== 0) {
        var error = cmd.stdout.toString();

        throw new Error(error);
      } else {
        var result = cmd.stdout.toString().split(eol).filter(function (v) {
          return v != '';
        }).concat(loaderSrc);

        debug('psc-package result: %o', result);

        CACHE_VAR.srcOption = result;

        return result;
      }
    } else {
      var _result = loaderOptions.src || [bowerPath, srcPath];

      CACHE_VAR.srcOption = _result;

      return _result;
    }
  }(loaderOptions.pscPackage);

  var options = Object.assign({
    context: webpackContext,
    psc: null,
    pscArgs: {},
    pscBundle: null,
    pscBundleArgs: {},
    pscIdeClient: null,
    pscIdeClientArgs: {},
    pscIdeServer: null,
    pscIdeServerArgs: {},
    pscIde: false,
    pscIdeColors: loaderOptions.psc === 'psa',
    pscPackage: false,
    bundleOutput: 'output/bundle.js',
    bundleNamespace: 'PS',
    bundle: false,
    warnings: true,
    watch: false,
    output: 'output',
    src: []
  }, loaderOptions, {
    src: srcOption
  });

  if (!CACHE_VAR.installed) {
    debugVerbose('installing purs-loader with options: %O', options);

    CACHE_VAR.installed = true;

    // invalidate loader CACHE_VAR when bundle is marked as invalid (in watch mode)
    this._compiler.plugin('invalid', function () {
      debugVerbose('invalidating loader CACHE_VAR');

      CACHE_VAR = {
        rebuild: options.pscIde,
        deferred: [],
        bundleModules: [],
        ideServer: CACHE_VAR.ideServer,
        psModuleMap: CACHE_VAR.psModuleMap,
        warnings: [],
        errors: [],
        compilationStarted: false,
        compilationFinished: false,
        compilationFailed: false,
        installed: CACHE_VAR.installed,
        srcOption: []
      };
    });

    // add psc warnings to webpack compilation warnings
    this._compiler.plugin('after-compile', function (compilation, callback) {
      CACHE_VAR.warnings.forEach(function (warning) {
        compilation.warnings.push(warning);
      });

      CACHE_VAR.errors.forEach(function (error) {
        compilation.errors.push(error);
      });

      callback();
    });
  }

  var psModuleName = PsModuleMap.matchModule(source);

  var psModule = {
    name: psModuleName,
    source: source,
    load: function load(_ref) {
      var js = _ref.js,
          map = _ref.map;
      return callback(null, js, map);
    },
    reject: function reject(error) {
      return callback(error);
    },
    srcPath: this.resourcePath,
    remainingRequest: loaderUtils.getRemainingRequest(this),
    srcDir: path.dirname(this.resourcePath),
    jsPath: path.resolve(path.join(options.output, psModuleName, 'index.js')),
    options: options,
    cache: CACHE_VAR,
    emitWarning: function emitWarning(warning) {
      if (options.warnings && warning.length) {
        CACHE_VAR.warnings.push(warning);
      }
    },
    emitError: pscMessage => {
      if (pscMessage.length) {

        CACHE_VAR.errors.push(pscMessage);

        const errors = splitPscErrors(pscMessage);
        for (const error of errors) {
          const [, relname] = matchLocation.exec(error) || [];
          if (!relname) continue;

          const filename = path.join(this.rootContext, relname);
          this.addDependency(filename);

          const [, fromModuleName] = matchModuleName.exec(error) || [];
          if (!fromModuleName) continue;

          for (const re of [matchMissingModuleName, matchMissingImportFromModuleName]) {
            const [, targetModuleName] = re.exec(error) || [];
            if (targetModuleName) {
              const resolved = resolveFilename({
                base: filename,
                from: fromModuleName,
                target: targetModuleName
              });
              this.addDependency(resolved);
            }
          }

          if (typeof this.extractDependencies === 'function') {
            for (const dep of this.extractDependencies(error)) {
              this.addDependency(dep);
            }
          }
        }
      }
    }
  };

  debug('loading %s', psModule.name);

  if (options.bundle) {
    CACHE_VAR.bundleModules.push(psModule.name);
  }

  if (CACHE_VAR.rebuild) {
    var connect = function connect() {
      if (!CACHE_VAR.ideServer) {
        CACHE_VAR.ideServer = true;

        return ide.connect(psModule).then(function (ideServer) {
          CACHE_VAR.ideServer = ideServer;
          return psModule;
        }).then(ide.loadWithRetry).catch(function (error) {
          if (CACHE_VAR.ideServer.kill) {
            debug('ide failed to initially load modules, stopping the ide server process');

            CACHE_VAR.ideServer.kill();
          }

          CACHE_VAR.ideServer = null;

          return Promise.reject(error);
        });
      } else {
        return Promise.resolve(psModule);
      }
    };

    var rebuild = function rebuild() {
      return ide.rebuild(psModule).then(function () {
        return toJavaScript(psModule).then(function (js) {
          return sourceMaps(psModule, js);
        }).then(psModule.load).catch(psModule.reject);
      }).catch(function (error) {
        if (error instanceof ide.UnknownModuleError) {
          // Store the modules that trigger a recompile due to an
          // unknown module error. We need to wait until compilation is
          // done before loading these files.

          CACHE_VAR.deferred.push(psModule);

          if (!CACHE_VAR.compilationStarted) {
            CACHE_VAR.compilationStarted = true;

            return compile(psModule).then(function () {
              CACHE_VAR.compilationFinished = true;
            }).then(function () {
              return Promise.map(CACHE_VAR.deferred, function (psModule) {
                return ide.load(psModule).then(function () {
                  return toJavaScript(psModule);
                }).then(function (js) {
                  return sourceMaps(psModule, js);
                }).then(psModule.load);
              });
            }).catch(function (error) {
              CACHE_VAR.compilationFailed = true;

              CACHE_VAR.deferred[0].reject(error);

              CACHE_VAR.deferred.slice(1).forEach(psModule => {
                psModule.reject(new Error('purs-loader failed'));
              })
            });
          } else if (CACHE_VAR.compilationFailed) {
            CACHE_VAR.deferred.pop().reject(new Error('purs-loader failed'));
          } else {
            // The compilation has started. We must wait until it is
            // done in order to ensure the module map contains all of
            // the unknown modules.
          }
        } else {
          debug('ide rebuild failed due to an unhandled error: %o', error);

          psModule.reject(error);
        }
      });
    };

    connect().then(rebuild);
  } else if (CACHE_VAR.compilationFinished) {
    debugVerbose('compilation is already finished, loading module %s', psModule.name);

    toJavaScript(psModule).then(function (js) {
      return sourceMaps(psModule, js);
    }).then(psModule.load).catch(psModule.reject);
  } else {
    // The compilation has not finished yet. We need to wait for
    // compilation to finish before the loaders run so that references
    // to compiled output are valid. Push the modules into the CACHE_VAR to
    // be loaded once the complation is complete.

    CACHE_VAR.deferred.push(psModule);

    if (!CACHE_VAR.compilationStarted) {
      CACHE_VAR.compilationStarted = true;

      compile(psModule).then(function () {
        CACHE_VAR.compilationFinished = true;
      }).then(function () {
        if (options.bundle) {
          return bundle(options, CACHE_VAR.bundleModules);
        }
      }).then(function () {
        return Promise.map(CACHE_VAR.deferred, function (psModule) {
          return toJavaScript(psModule).then(function (js) {
            return sourceMaps(psModule, js);
          }).then(psModule.load);
        });
      }).catch(function (error) {


        CACHE_VAR.compilationFailed = true;

        CACHE_VAR.deferred[0].reject(error);

        CACHE_VAR.deferred.slice(1).forEach(psModule => {
          psModule.reject(new Error('purs-loader failed'));
        })
      });
    } else if (CACHE_VAR.compilationFailed) {
      CACHE_VAR.deferred.pop().reject(new Error('purs-loader failed'));
    } else {
      // The complation has started. Nothing to do but wait until it is
      // done before loading all of the modules.
    }
  }
};
